File: py/aoc2024/day15.py
"""
Day 15: Warehouse Woes
"""

SAMPLE_INPUT_1 = """
##########
#..O..O.O#
#......O.#
#.OO..O.O#
#..O@..O.#
#O#..O...#
#O..O..O.#
#.OO.O.OO#
#....O...#
##########

<vv>^<v^>v>^vv^v>v<>v^v<v<^vv<<<^><<><>>v<vvv<>^v^>^<<<><<v<<<v^vv^v>^
vvv<<^>^v^^><<>>><>^<<><^vv^^<>vvv<>><^^v>^>vv<>v<<<<v<^v>^<^^>>>^<v<v
><>vv>v^v^<>><>>>><^^>vv>v<^^^>>v^v^<^^>v^^>v^<^v>v<>>v^v^<v>v^^<^^vv<
<<v<^>>^^^^>>>v^<>vvv^><v<<<>^^^vv^<vvv>^>v<^^^^v<>^>vvvv><>>v^<<^^^^^
^><^><>>><>^^<<^^v>>><^<v>^<vv>>v>>>^v><>^v><<<<v>>v<v<v>vvv>^<><<>^><
^>><>^v<><^vvv<^^<><v<<<<<><^v<<<><<<^^<v<^^^><^>>^<v^><<<^>>^v<v^v<v^
>^>>^v>vv>^<<^v<>><<><<v<<v><>v<^vv<<<>^^v^>^^>>><<^v>>v^v><^^>>^<>vv^
<><^^>^^^<><vvvvv^v<v<<>^v<v>v<<^><<><<><<<^^<<<^<<>><<><^^^>^^<>^>v<>
^^>vv<^v^v<vv>^<><v<^v>^^^>>>^^vvv^>vvv<>>>^<^>>>>>^<<^v>^vvv<>^<><<v>
v^^>>><<^^<>>^v^<v^vv<>v^<<>^<^v^v><^<<<><<^<v><v<>vv>>v><v^<vv<>v^<<^
"""
SAMPLE_INPUT_2 = """
########
#..O.O.#
##@.O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

<^^>>>vv<v>>v<<
"""
SAMPLE_INPUT_3 = """
#######
#...#.#
#.....#
#..OO@#
#..O..#
#.....#
#######

<vv<<^^<<^^
"""


class _Warehouse:
    def __init__(self, grid: str):
        self.grid = [list(line) for line in grid.strip("\n").splitlines()]
        for y, line in enumerate(self.grid):
            for x, char in enumerate(line):
                if char == "@":
                    self.pos = y, x

    def __str__(self) -> str:
        return "\n".join("".join(line) for line in self.grid)

    def double(self):
        return _Warehouse(
            "\n".join(
                "".join("@." if c == "@" else "[]" if c == "O" else c + c for c in line)
                for line in self.grid
            )
        )

    def move(self, char):
        match char:
            case "^":
                dy, dx = -1, 0
            case "v":
                dy, dx = 1, 0
            case "<":
                dy, dx = 0, -1
            case ">":
                dy, dx = 0, 1
            case _:
                return
        levels = []
        front = {self.pos}
        while front:
            levels.append(front)
            front = {(y + dy, x + dx) for y, x in front}
            for y, x in list(front):
                match self.grid[y][x]:
                    case ".":
                        front.remove((y, x))
                    case "[" if dy:
                        front.add((y, x + 1))
                    case "]" if dy:
                        front.add((y, x - 1))
                    case "O" | "[" | "]":
                        pass
                    case _:
                        return
        for front in levels[::-1]:
            for y, x in front:
                self.grid[y + dy][x + dx] = self.grid[y][x]
                self.grid[y][x] = "."
        self.pos = self.pos[0] + dy, self.pos[1] + dx

    def gps(self) -> int:
        return sum(
            100 * y + x
            for y, line in enumerate(self.grid)
            for x, char in enumerate(line)
            if char == "O" or char == "["
        )


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT_2)
    2028
    >>> part1(SAMPLE_INPUT_1)
    10092
    """
    warehouse, moves = data.split("\n\n", maxsplit=1)
    warehouse = _Warehouse(warehouse)
    for move in moves:
        warehouse.move(move)
    return warehouse.gps()


def part2(data: str) -> int:
    """
    >>> part2(SAMPLE_INPUT_3)
    618
    >>> part2(SAMPLE_INPUT_1)
    9021
    """
    warehouse, moves = data.split("\n\n", maxsplit=1)
    warehouse = _Warehouse(warehouse).double()
    for move in moves:
        warehouse.move(move)
    return warehouse.gps()


parts = (part1, part2)

File: rs/src/day15.rs
use std::collections::BTreeSet;
use std::ops::IndexMut;

type YX = (usize, usize);

fn parse(data: &str) -> Option<(Vec<Vec<u8>>, YX, &str)> {
    let (map, moves) = data.split_once("\n\n").unwrap_or((data, ""));
    let map = map
        .lines()
        .map(|line| line.bytes().collect())
        .collect::<Vec<Vec<_>>>();
    let robot = map.iter().enumerate().find_map(|(y, line)| {
        line.iter()
            .enumerate()
            .find(|(_, b)| **b == b'@')
            .map(|(x, _)| (y, x))
    });
    Some((map, robot?, moves))
}

fn move_robot<Line>(map: &mut [Line], robot: YX, c: char) -> Option<YX>
where
    Line: IndexMut<usize, Output = u8>,
{
    let (dy, dx) = match c {
        '^' => (usize::MAX, 0),
        'v' => (1, 0),
        '<' => (0, usize::MAX),
        '>' => (0, 1),
        _ => return None,
    };
    let mut levels = vec![];
    let mut front: BTreeSet<_> = [robot].into();
    while !front.is_empty() {
        let back = front;
        front = BTreeSet::new();
        for (y, x) in &back {
            let (y, x) = (y.wrapping_add(dy), x.wrapping_add(dx));
            match map[y][x] {
                b'.' => continue,
                b'O' => {
                    front.insert((y, x));
                }
                b'[' => {
                    front.insert((y, x));
                    if dy != 0 {
                        front.insert((y, x + 1));
                    }
                }
                b']' => {
                    front.insert((y, x));
                    if dy != 0 {
                        front.insert((y, x - 1));
                    }
                }
                _ => return None,
            }
        }
        levels.push(back);
    }
    for front in levels.into_iter().rev() {
        for (y, x) in front {
            map[y.wrapping_add(dy)][x.wrapping_add(dx)] = map[y][x];
            map[y][x] = b'.';
        }
    }
    Some((robot.0.wrapping_add(dy), robot.1.wrapping_add(dx)))
}

pub fn part1(data: &str) -> Option<usize> {
    let (mut map, mut robot, moves) = parse(data)?;
    for c in moves.chars() {
        robot = move_robot(&mut map, robot, c).unwrap_or(robot);
    }
    Some(
        map.into_iter()
            .enumerate()
            .flat_map(|(y, line)| {
                line.into_iter().enumerate().filter_map(move |(x, b)| {
                    if b == b'O' {
                        Some(100 * y + x)
                    } else {
                        None
                    }
                })
            })
            .sum(),
    )
}

pub fn part2(data: &str) -> Option<usize> {
    let (mut map, mut robot, moves) = parse(data)?;
    for line in &mut map {
        *line = line
            .iter()
            .flat_map(|&b| match b {
                b'@' => *b"@.",
                b'O' => *b"[]",
                _ => [b, b],
            })
            .collect();
    }
    robot.1 *= 2;
    for c in moves.chars() {
        robot = move_robot(&mut map, robot, c).unwrap_or(robot);
    }
    Some(
        map.into_iter()
            .enumerate()
            .flat_map(|(y, line)| {
                line.into_iter().enumerate().filter_map(move |(x, b)| {
                    if b == b'[' {
                        Some(100 * y + x)
                    } else {
                        None
                    }
                })
            })
            .sum(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE_1: &str = indoc! {"
        ##########
        #..O..O.O#
        #......O.#
        #.OO..O.O#
        #..O@..O.#
        #O#..O...#
        #O..O..O.#
        #.OO.O.OO#
        #....O...#
        ##########

        <vv>^<v^>v>^vv^v>v<>v^v<v<^vv<<<^><<><>>v<vvv<>^v^>^<<<><<v<<<v^vv^v>^
        vvv<<^>^v^^><<>>><>^<<><^vv^^<>vvv<>><^^v>^>vv<>v<<<<v<^v>^<^^>>>^<v<v
        ><>vv>v^v^<>><>>>><^^>vv>v<^^^>>v^v^<^^>v^^>v^<^v>v<>>v^v^<v>v^^<^^vv<
        <<v<^>>^^^^>>>v^<>vvv^><v<<<>^^^vv^<vvv>^>v<^^^^v<>^>vvvv><>>v^<<^^^^^
        ^><^><>>><>^^<<^^v>>><^<v>^<vv>>v>>>^v><>^v><<<<v>>v<v<v>vvv>^<><<>^><
        ^>><>^v<><^vvv<^^<><v<<<<<><^v<<<><<<^^<v<^^^><^>>^<v^><<<^>>^v<v^v<v^
        >^>>^v>vv>^<<^v<>><<><<v<<v><>v<^vv<<<>^^v^>^^>>><<^v>>v^v><^^>>^<>vv^
        <><^^>^^^<><vvvvv^v<v<<>^v<v>v<<^><<><<><<<^^<<<^<<>><<><^^^>^^<>^>v<>
        ^^>vv<^v^v<vv>^<><v<^v>^^^>>>^^vvv^>vvv<>>>^<^>>>>>^<<^v>^vvv<>^<><<v>
        v^^>>><<^^<>>^v^<v^vv<>v^<<>^<^v^v><^<<<><<^<v><v<>vv>>v><v^<vv<>v^<<^
    "};

    static EXAMPLE_2: &str = indoc! {"
        ########
        #..O.O.#
        ##@.O..#
        #...O..#
        #.#.O..#
        #...O..#
        #......#
        ########

        <^^>>>vv<v>>v<<
    "};

    static EXAMPLE_3: &str = indoc! {"
        #######
        #...#.#
        #.....#
        #..OO@#
        #..O..#
        #.....#
        #######

        <vv<<^^<<^^
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(Some(2028), part1(EXAMPLE_2));
        assert_eq!(Some(10092), part1(EXAMPLE_1));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(Some(618), part2(EXAMPLE_3));
        assert_eq!(Some(9021), part2(EXAMPLE_1));
    }
}
